### PR TITLE
few missing allocation failure checks and releases on error paths

### PR DESCRIPTION
**Pull request by ajmohan**
- missing checks for allocation failure.
- releasing memory in few missing error paths

### CODE REVIEW

**Comment by richsalz** [comment]
+1

File: crypto/ec/ec_kmeth.c [hunk 1]
```diff
@@ -168,6 +168,8 @@ EC_KEY *EC_KEY_new_method(ENGINE *engine)
     ret->references = 1;
 
     if (ret->meth->init != NULL && ret->meth->init(ret) == 0) {
+        CRYPTO_free_ex_data(CRYPTO_EX_INDEX_EC_KEY, ret, &ret->ex_data);
+        CRYPTO_THREAD_lock_free(ret->lock);
```
Comment (mattcaswell) [inline crypto/ec/ec_kmeth.c]: Do we need to call `ENGINE_finish()` (guarded by OPENSSL_NO_ENGINE) like the other failure case above? It seems to me there is a lot of duplicated clean up code in this function which could be pulled into a separate `goto` block at the end.

File: crypto/evp/bio_b64.c [hunk 2]
```diff
@@ -116,16 +116,20 @@ static int b64_new(BIO *bi)
...
     ctx = OPENSSL_zalloc(sizeof(*ctx));
     if (ctx == NULL)
-        return (0);
+        return 0;
 
     ctx->cont = 1;
     ctx->start = 1;
     ctx->base64 = EVP_ENCODE_CTX_new();
+    if (ctx->base64 == NULL) {
+        OPENSSL_free (ctx);
```
Comment (mattcaswell) [inline crypto/evp/bio_b64.c]: No space between the function name and the parens please.

File: crypto/evp/bio_ok.c [hunk 3]
```diff
@@ -178,15 +178,19 @@ static int ok_new(BIO *bi)
...
     ctx = OPENSSL_zalloc(sizeof(*ctx));
     if (ctx == NULL)
-        return (0);
+        return 0;
 
     ctx->cont = 1;
     ctx->sigio = 1;
     ctx->md = EVP_MD_CTX_new();
+    if (ctx->md == NULL) {
+        OPENSSL_free (ctx);
```
Comment (mattcaswell) [inline crypto/evp/bio_ok.c]: Please remove space here too.

File: crypto/ec/ec_kmeth.c (hunk 1 above)
Comment (ajmohan) [inline crypto/ec/ec_kmeth.c]: Yes, ENGINE_finish() is required, done. 
As you mentioned cleanup code can be moved to a **goto** block.

File: crypto/ec/ec_kmeth.c [hunk 4]
```diff
@@ -166,6 +166,12 @@ EC_KEY *EC_KEY_new_method(ENGINE *engine)
     ret->references = 1;
 
     if (ret->meth->init != NULL && ret->meth->init(ret) == 0) {
+        ECerr(EC_F_EC_KEY_NEW_METHOD, ERR_R_INIT_FAIL);
+#ifndef OPENSSL_NO_ENGINE
+        ENGINE_finish(ret->engine);
```
Comment (FdaSilvaYY) [inline crypto/ec/ec_kmeth.c]: This change is redoing the work of  EC_KEY_free().
So it creates a double-free issue, three times.
---
IMHO, this method could be simplied as I do in another PR 
See  e178c76fe9e4a197b9dab931451c57c340a25a8e

Comment (ajmohan) [inline crypto/ec/ec_kmeth.c]: Yes, you are right. EC_KEY_free() is already doing it.

**Comment by levitte** [comment]
Sorry it's taken such time to get to this.  It needs to be rebased on top of a fresher master.

**Comment by mattcaswell** [comment]
Ping @ajmohan, @levitte is correct. This needs a rebase before we can take it.

**Comment by FdaSilvaYY** [comment]
FYI, half of these changes are already merged, through various other commits.
That's why there is conflicts.

**Comment by ajmohan** [comment]
I am moving the remaining changes to a new PR and closing this PR. Thanks

**Comment by ajmohan** [comment]
Created PR #1031
